### PR TITLE
fix bug in normal random number generator

### PR DESCRIPTION
**Pull request by darrenjw**
There is a slight bug in the normal random number generation function "randn" in the prelude that occasionally leads to the generation of +/-Infinity. This can cause mysterious failures of large stochastic simulation codes involving normal random number generation. This PR contains a one-line fix which seems to resolve the problem.

Prior to the fix:

z: (Fin 100000000=>Float) = randn_vec (new_key 42)
maximum z
> Infinity
minimum z
> -Infinity

After applying the fix:

z: (Fin 100000000=>Float) = randn_vec (new_key 42)
maximum z
> 5.559744
minimum z
> -5.593339

### CODE REVIEW

File: lib/prelude.dx [hunk 1]
```diff
@@ -1903,7 +1903,7 @@ def rand_mat {a} (n:Nat) (m:Nat) (f: Key -> a) (k: Key) : Fin n => Fin m => a =
 
 def randn (k:Key) : Float =
   [k1, k2] = split_key k
-  u1 = rand k1
+  u1 = 1.0 - (rand k1)
```
Comment (axch) [inline lib/prelude.dx:1906]:
```suggestion
  -- rand is uniform between 0 and 1, but implemented such that it rounds to 0
  -- (in float32) once every few million draws, but never rounds to 1.
  u1 = 1.0 - (rand k1)
```

**Review by axch — CHANGES_REQUESTED** [review]
Thanks for the fix!

I approved running our CI tests; could you fix the couple of test changes?  (They're hardcoded sampling results, so unfortunately we have to keep updating them whenever we update the samplers; but at least there aren't too many.)

I also suggested a code comment to add to the implementation to explain why it is the way it is.  We'd certainly be happy with better wording if you don't like my suggestion, but I think there ought to be something there.

**Review by apaszke — APPROVED** [review]
Actually, I think that rand generates values in the interval [0, 1), as is conventional for many uniform PRNGs. This LGTM, thanks for the PR!

My main concern is whether the rounding from `1 - x` will not bias the sampler unexpectedly (floating point numbers are denser around 0 than around 1), but I don't think that's the case with our implementation.

**Comment by darrenjw** [comment]
The "1-u" fix is used quite often, so I think it should be OK.

There are still a couple of example files with failing tests, but they don't obviously use "randn" anywhere, so I'm not sure if those failures are expected?

**Comment by apaszke** [comment]
I think those are expected. randn might be used indirectly through the [`Arbitrary` instance for floats](https://github.com/google-research/dex-lang/blob/main/lib/prelude.dx#L1964).

**Comment by apaszke** [comment]
That is, please update them, and let's merge it!

**Review by axch — APPROVED** [review]

**Pull request merged**